File: synth/board.py
"""
FigJam board output. Turns a run's insights into (a) a layout: one section per
insight holding a claim sticky, receipt stickies, counter-evidence stickies,
and an opportunity sticky, with connectors from claim to counter-evidence;
and (b) one `use_figma` script per section that a host holding Figma
credentials can execute against a board. Motif never talks to Figma itself.

Colours are FigJam sticky palette values in hex/255 notation (step zero showed
that rounded decimals and some palette entries round-trip as CUSTOM).
"""

from __future__ import annotations

import json

from synth.agents import turn_ids

# Sticky palette (from Figma's figma-use-figjam reference). Roles -> hex.
STICKY = {
    "claim_high": "B3EFBD",    # green
    "claim_medium": "FFE299",  # yellow
    "claim_low": "FFD3A8",     # orange
    "receipt": "FFFFFF",       # white
    "counter": "FFA8DB",       # pink
    "opportunity": "A8DAFF",   # blue
    "contested": "D3BDFF",     # violet
}
SECTION_FILL = "F5FBFF"        # light blue
CONNECTOR = "F849C1"           # pink stroke

SECTION_W = 1600
SLOT_H = 1200          # tallest section seen on the R5 board was 984px
GAP = 120
PAD = 48
SPACING = 40


def _q(item) -> tuple[str, str]:
    if isinstance(item, dict):
        return item.get("turn", ""), item.get("quote", "") or ""
    return str(item), ""


def layout(insights: list[dict], meta: dict | None = None, columns: int = 2,
           origin: tuple[float, float] = (0, 0)) -> dict:
    """Pure data: what goes on the board and where each section sits."""
    meta = meta or {}
    sections = []
    for n, ins in enumerate(insights):
        conf = (ins.get("confidence") or "low").lower()
        stickies = [{"key": "claim", "role": "claim", "wide": True,
                     "color": STICKY.get(f"claim_{conf}", STICKY["claim_low"]),
                     "text": f"{ins.get('id', '?')} · {conf.upper()}\n{ins.get('title', '')}\n\n{ins.get('claim', '')}"}]
        for i, item in enumerate(ins.get("evidence") or []):
            turn, quote = _q(item)
            stickies.append({"key": f"receipt{i}", "role": "receipt", "wide": False, "color": STICKY["receipt"],
                             "text": f"{turn}\n“{quote}”" if quote else f"{turn}\n(no receipt)"})
        counters = []
        for i, item in enumerate(ins.get("counter_evidence") or []):
            turn, quote = _q(item)
            key = f"counter{i}"
            counters.append(key)
            note = f"\n{ins.get('counter_note')}" if i == 0 and ins.get("counter_note") else ""
            stickies.append({"key": key, "role": "counter", "wide": False, "color": STICKY["counter"],
                             "text": f"COUNTER {turn}\n“{quote}”{note}" if quote else f"COUNTER {turn}{note}"})
        if ins.get("opportunity"):
            stickies.append({"key": "opportunity", "role": "opportunity", "wide": True,
                             "color": STICKY["opportunity"], "text": f"OPPORTUNITY\n{ins['opportunity']}"})
        if ins.get("critic_flags"):
            stickies.append({"key": "contested", "role": "contested", "wide": True, "color": STICKY["contested"],
                             "text": "CONTESTED — critic's unresolved objections:\n" + "\n".join(ins["critic_flags"])})
        sections.append({
            "insight_id": ins.get("id", f"I-{n + 1:02d}"),
            "name": f"{ins.get('id', '?')} · {ins.get('title', '')}"[:80],
            "x": origin[0] + (n % columns) * (SECTION_W + GAP),
            "y": origin[1] + (n // columns) * (SLOT_H + GAP),
            "width": SECTION_W,
            "fill": SECTION_FILL,
            "stickies": stickies,
            "connectors": [{"from": "claim", "to": c, "label": "contested by"} for c in counters],
            "sources": ins.get("sources", []),
            "turns": turn_ids(ins.get("evidence")) + turn_ids(ins.get("counter_evidence")),
        })
    return {
        "title": f"Motif synthesis — run {meta.get('run_id', '')}".strip(" —"),
        "kind": "synthesis",
        "run_id": meta.get("run_id"),
        "columns": columns,
        "origin": list(origin),
        "n_sections": len(sections),
        "sections": sections,
        "card": _card(meta, {
            "insights": len(insights),
            "contested": [i.get("id") for i in insights if i.get("critic_flags")],
        }),
    }


def _card(meta: dict, counts: dict) -> dict:
    """What a board's run card says: the run, the question, the numbers, and any open objections."""
    cfg = meta.get("config") or {}
    corpus = meta.get("corpus") or {}
    return {
        "run_id": meta.get("run_id"),
        "question": cfg.get("question"),
        "transcripts": corpus.get("transcripts") or cfg.get("transcripts") or [],
        "words": corpus.get("words"),
        "condition": meta.get("condition"),
        "iterations": meta.get("iterations"),
        "stop_reason": meta.get("stop_reason"),
        "cost": meta.get("cost"),
        "wall_seconds": meta.get("wall_seconds"),
        "started": meta.get("started"),
        **counts,
    }


VERDICT_STICKY = {"claim_pass": STICKY["claim_high"], "claim_fail": STICKY["counter"],
                  "claim_warn": STICKY["claim_medium"], "fail": STICKY["counter"],
                  "warn": STICKY["claim_medium"], "theme": STICKY["contested"]}


def verdict_layout(insights: list[dict], verdict: dict, meta: dict | None = None, columns: int = 2,
                   origin: tuple[float, float] = (0, 0)) -> dict:
    """A critic verdict as a board: one section per checked claim, coloured by outcome, with one
    sticky per objection; a final section for corpus-level objections (insight_id '*')."""
    meta = meta or {}
    failures = verdict.get("failures") or []
    by_id: dict = {}
    for f in failures:
        by_id.setdefault(f.get("insight_id") or "*", []).append(f)
    sections = []

    def place(n: int) -> tuple[float, float]:
        return origin[0] + (n % columns) * (SECTION_W + GAP), origin[1] + (n // columns) * (SLOT_H + GAP)

    for n, ins in enumerate(insights):
        objections = by_id.pop(ins.get("id"), [])
        worst = "fail" if any(f.get("severity", "fail") == "fail" for f in objections) else \
                "warn" if objections else "pass"
        stickies = [{"key": "claim", "role": "claim", "wide": True, "row": 1,
                     "color": VERDICT_STICKY[f"claim_{worst}"],
                     "text": f"{ins.get('id', '?')} · {worst.upper()}\n{ins.get('title', '')}\n\n{ins.get('claim', '')}"}]
        for i, f in enumerate(objections):
            sev = f.get("severity", "fail")
            turns = ", ".join(f.get("turns") or [])
            stickies.append({"key": f"objection{i}", "role": sev, "wide": True, "row": 2,
                             "color": VERDICT_STICKY[sev],
                             "text": f"{sev.upper()} · {f.get('rule')}\n{f.get('detail', '')}" + (f"\n\nturns: {turns}" if turns else "")})
        x, y = place(n)
        sections.append({"insight_id": ins.get("id", f"I-{n + 1:02d}"),
                         "name": f"{ins.get('id', '?')} · {worst} · {ins.get('title', '')}"[:80],
                         "x": x, "y": y, "width": SECTION_W, "fill": SECTION_FILL,
                         "stickies": stickies, "connectors": [], "sources": ins.get("sources", []),
                         "turns": turn_ids(ins.get("evidence")) + turn_ids(ins.get("counter_evidence"))})
    rest = [f for fs in by_id.values() for f in fs]
    if rest:
        stickies = [{"key": f"corpus{i}", "role": "theme", "wide": True, "row": 1, "color": VERDICT_STICKY["theme"],
                     "text": f"{f.get('rule', '').upper()} (whole corpus)\n{f.get('detail', '')}"
                             + (f"\n\nturns: {', '.join(f.get('turns') or [])}" if f.get("turns") else "")}
                    for i, f in enumerate(rest)]
        x, y = place(len(sections))
        sections.append({"insight_id": "*", "name": "Objections about the whole corpus", "x": x, "y": y,
                         "width": SECTION_W, "fill": SECTION_FILL, "stickies": stickies, "connectors": [],
                         "sources": [], "turns": []})
    n_fail = sum(1 for f in failures if f.get("severity", "fail") == "fail")
    n_warn = sum(1 for f in failures if f.get("severity") == "warn")
    return {
        "title": f"Motif critique — run {meta.get('run_id', '')}".strip(" —"),
        "kind": "verdict",
        "run_id": meta.get("run_id"),
        "columns": columns,
        "origin": list(origin),
        "n_sections": len(sections),
        "sections": sections,
        "card": _card(meta, {"claims": len(insights), "pass": bool(verdict.get("pass")), "fails": n_fail,
                             "warns": n_warn, "skipped_rules": verdict.get("skipped_rules") or []}),
    }


SCRIPT_TEMPLATE = """\
// Motif board section {insight_id} — generated; run with use_figma (skills: figma-use, figma-use-figjam)
const S = {spec};
const h = (hex) => ({{ r: parseInt(hex.slice(0, 2), 16) / 255, g: parseInt(hex.slice(2, 4), 16) / 255, b: parseInt(hex.slice(4, 6), 16) / 255 }});
const section = figma.createSection();
section.name = S.name;
section.fills = [{{ type: 'SOLID', color: h(S.fill) }}];
section.x = S.x; section.y = S.y;
const probe = figma.createSticky();
await figma.loadFontAsync(probe.text.fontName);
probe.remove();
const made = {{}};
for (const st of S.stickies) {{
  const s = figma.createSticky();
  s.text.characters = st.text;
  s.isWideWidth = !!st.wide;
  s.fills = [{{ type: 'SOLID', color: h(st.color) }}];
  s.authorVisible = false;
  section.appendChild(s);
  made[st.key] = s;
}}
// Row 1: claim then receipts; row 2: counters then opportunity/contested. Two-pass: measure, then place.
const rowOf = (s) => s.row || ((s.role === 'claim' || s.role === 'receipt') ? 1 : 2);
const row1 = S.stickies.filter(s => rowOf(s) === 1).map(s => made[s.key]);
const row2 = S.stickies.filter(s => rowOf(s) === 2).map(s => made[s.key]);
let y = {pad}, maxRight = 0;
for (const row of [row1, row2]) {{
  if (!row.length) continue;
  let x = {pad};
  for (const s of row) {{ s.x = x; s.y = y; x += s.width + {spacing}; }}
  maxRight = Math.max(maxRight, x - {spacing});
  y += Math.max(...row.map(s => s.height)) + {spacing};
}}
section.resize(Math.max(S.width, maxRight + {pad}), y - {spacing} + {pad});
const font = {{ family: 'Inter', style: 'Medium' }};
await figma.loadFontAsync(font);
const connectors = [];
for (const c of S.connectors) {{
  const conn = figma.createConnector();
  conn.connectorStart = {{ endpointNodeId: made[c.from].id, magnet: 'AUTO' }};
  conn.connectorEnd = {{ endpointNodeId: made[c.to].id, magnet: 'AUTO' }};
  conn.connectorStartStrokeCap = 'NONE';
  conn.connectorEndStrokeCap = 'ARROW_LINES';
  conn.strokes = [{{ type: 'SOLID', color: h('{connector}') }}];
  conn.text.fontName = font;
  conn.text.characters = c.label;
  connectors.push(conn.id);
}}
return {{
  insightId: S.insight_id,
  sectionId: section.id,
  stickyIds: Object.fromEntries(Object.entries(made).map(([k, s]) => [k, s.id])),
  connectorIds: connectors,
  createdNodeIds: [section.id, ...Object.values(made).map(s => s.id), ...connectors],
  sectionSize: [section.width, section.height],
}};
"""


def scripts(board: dict) -> list[dict]:
    """One use_figma script per section. Each is self-contained and returns the node ids it made."""
    out = []
    for sec in board["sections"]:
        spec = {k: sec[k] for k in ("insight_id", "name", "x", "y", "width", "fill", "stickies", "connectors")}
        code = SCRIPT_TEMPLATE.format(insight_id=sec["insight_id"], spec=json.dumps(spec, ensure_ascii=False),
                                      pad=PAD, spacing=SPACING, connector=CONNECTOR)
        out.append({"insight_id": sec["insight_id"], "code": code, "chars": len(code)})
    return out


HOST_INSTRUCTIONS = (
    "Run each script in order with Figma's `use_figma` tool against a FigJam board (URL "
    "figma.com/board/<fileKey>/...), passing skillNames 'figma-use,figma-use-figjam' "
    "(prefix 'resource:' if loaded as MCP resources). Each script creates one section and "
    "returns the node ids it made; every script must return ids, and a script that returns "
    "no ids has failed. Read the board back with `get_figjam` afterwards to confirm."
)

File: surfaces/figma/src/board.ts
/**
 * Board writer: a Motif layout (synth/board.py::layout or ::verdict_layout, fetched from the
 * hosted engine) onto the current page. Runs on the plugin main thread; the same function is
 * bundled standalone (test/board_script.mjs) and executed through Figma's MCP `use_figma`, so the
 * renderer that ships is the renderer that was tested on a real board.
 *
 * FigJam: one section per insight, stickies in two rows (claim + receipts; counters + opportunity +
 * contested), "contested by" connectors from the claim to each counter-evidence sticky, and a run
 * card above the grid. Mirrors the `use_figma` script Motif's MCP `motif_board` tool emits.
 * Figma Design: the same sections and positions; a sticky becomes an auto-layout frame with a
 * text child (no sticky node in Design), and connectors are dropped (no connector node in Design),
 * so counter frames keep their "COUNTER" caption as the link.
 *
 * Every sticky must exist when the function returns; a missing one is a thrown error, never a
 * partial success. Layout coordinates are absolute from the layout's origin; `origin` re-bases them.
 */

export interface Sticky { key: string; role: string; wide: boolean; color: string; text: string; row?: number }
export interface Connector { from: string; to: string; label: string }
export interface Section {
  insight_id: string; name: string; x: number; y: number; width: number; fill: string;
  stickies: Sticky[]; connectors: Connector[];
}
export interface Card {
  run_id?: string | null; question?: string | null; transcripts?: string[]; words?: number | null;
  condition?: string | null; iterations?: number | null; stop_reason?: string | null; cost?: number | null;
  wall_seconds?: number | null; started?: string | null;
  insights?: number; contested?: string[]; claims?: number; pass?: boolean; fails?: number; warns?: number;
  skipped_rules?: string[];
}
export interface Layout {
  title: string; kind?: "synthesis" | "verdict"; run_id?: string | null; columns: number;
  origin: [number, number]; n_sections: number; sections: Section[]; card?: Card;
}
export interface BoardResult {
  sectionIds: string[]; stickyIds: string[]; connectorIds: string[]; cardId: string | null; nodeIds: string[];
  counts: { sections: number; stickies: number; connectors: number };
  bounds: { x: number; y: number; width: number; height: number };
  editor: string;
}

export const PAD = 48;
export const SPACING = 40;
export const CONNECTOR = "F849C1";
export const CARD_GAP = 120;
export const CARD_COLORS = { ok: "B3EFBD", attention: "D3BDFF" };

const hex = (h: string): RGB => ({
  r: parseInt(h.slice(0, 2), 16) / 255, g: parseInt(h.slice(2, 4), 16) / 255, b: parseInt(h.slice(4, 6), 16) / 255,
});

/** 11482 -> "11,482" without relying on the sandbox's locale (Figma's main thread formats bare). */
export function thousands(n: number): string {
  return String(Math.round(n)).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

function rowOf(s: Sticky): number {
  return s.row ?? (s.role === "claim" || s.role === "receipt" ? 1 : 2);
}

/** A free spot to the right of everything on the page, or the page origin when it is empty. */
export function chooseOrigin(page: PageNode): [number, number] {
  const kids = page.children;
  if (!kids.length) return [0, 0];
  let maxX = -Infinity, minY = Infinity;
  for (const n of kids) {
    if (!("x" in n) || !("width" in n)) continue;
    maxX = Math.max(maxX, n.x + n.width);
    minY = Math.min(minY, n.y);
  }
  if (!isFinite(maxX)) return [0, 0];
  return [Math.ceil((maxX + 400) / 100) * 100, Math.floor(minY / 100) * 100];
}

export function cardText(layout: Layout): string {
  const c = layout.card ?? {};
  const mins = c.wall_seconds != null ? `${(c.wall_seconds / 60).toFixed(1)} min` : "";
  const cost = c.cost != null ? `$${c.cost.toFixed(2)}` : "";
  const corpus = `${(c.transcripts ?? []).length} transcript${(c.transcripts ?? []).length === 1 ? "" : "s"}`
    + ((c.transcripts ?? []).length ? ` (${(c.transcripts ?? []).join(", ")})` : "")
    + (c.words != null ? `, ${thousands(c.words)} words` : "");
  const lines = [layout.kind === "verdict" ? `MOTIF CRITIQUE · run ${c.run_id ?? layout.run_id ?? "?"}` : `MOTIF SYNTHESIS · run ${c.run_id ?? layout.run_id ?? "?"}`];
  if (c.question) lines.push(`Question: ${c.question}`);
  lines.push(corpus);
  if (layout.kind === "verdict") {
    lines.push(`${c.pass ? "PASS" : "FAIL"}: ${c.fails ?? 0} fail(s), ${c.warns ?? 0} warning(s) on ${c.claims ?? layout.n_sections} claim(s)`);
    if (c.skipped_rules?.length) lines.push(`Not checked: ${c.skipped_rules.join(", ")}`);
  } else {
    lines.push(`${c.insights ?? layout.n_sections} insights` + (c.contested?.length ? `, ${c.contested.length} contested: ${c.contested.join(", ")}` : ", none contested"));
    lines.push([c.condition ? `Condition ${c.condition}` : "", c.iterations != null ? `${c.iterations} round(s)` : "",
      c.stop_reason ? `stop=${c.stop_reason}` : "", cost, mins].filter(Boolean).join(", "));
  }
  lines.push("Silence is never agreement: read contested items with the objection in view. Every receipt is verbatim; check any citation against the transcript.");
  if (c.started) lines.push(String(c.started).slice(0, 19).replace("T", " "));
  return lines.join("\n");
}

function cardColor(layout: Layout): string {
  const c = layout.card ?? {};
  const attention = layout.kind === "verdict" ? !c.pass : (c.contested?.length ?? 0) > 0;
  return attention ? CARD_COLORS.attention : CARD_COLORS.ok;
}

// ----------------------------------------------------------------- FigJam

async function renderFigJam(layout: Layout, dx: number, dy: number): Promise<BoardResult> {
  const probe = figma.createSticky();
  await figma.loadFontAsync(probe.text.fontName as FontName);
  probe.remove();
  const labelFont: FontName = { family: "Inter", style: "Medium" };
  await figma.loadFontAsync(labelFont);

  const sectionIds: string[] = [], stickyIds: string[] = [], connectorIds: string[] = [];
  const all: SceneNode[] = [];

  const makeSection = (name: string, fill: string, x: number, y: number): SectionNode => {
    const section = figma.createSection();
    section.name = name;
    section.fills = [{ type: "SOLID", color: hex(fill) }];
    section.x = x; section.y = y;
    return section;
  };
  const makeSticky = (section: SectionNode, st: Sticky): StickyNode => {
    const s = figma.createSticky();
    s.text.characters = st.text;
    s.isWideWidth = !!st.wide;
    s.fills = [{ type: "SOLID", color: hex(st.color) }];
    s.authorVisible = false;
    section.appendChild(s);
    return s;
  };

  for (const sec of layout.sections) {
    const section = makeSection(sec.name, sec.fill, sec.x + dx, sec.y + dy);
    const made: Record<string, StickyNode> = {};
    for (const st of sec.stickies) made[st.key] = makeSticky(section, st);
    if (Object.keys(made).length !== sec.stickies.length) {
      throw new Error(`section ${sec.insight_id}: ${Object.keys(made).length} of ${sec.stickies.length} stickies created`);
    }
    const rows = [1, 2].map((r) => sec.stickies.filter((s) => rowOf(s) === r).map((s) => made[s.key]));
    let y = PAD, maxRight = 0;
    for (const row of rows) {
      if (!row.length) continue;
      let x = PAD;
      for (const s of row) { s.x = x; s.y = y; x += s.width + SPACING; }
      maxRight = Math.max(maxRight, x - SPACING);
      y += Math.max(...row.map((s) => s.height)) + SPACING;
    }
    section.resizeWithoutConstraints(Math.max(sec.width, maxRight + PAD), y - SPACING + PAD);
    for (const c of sec.connectors) {
      const conn = figma.createConnector();
      conn.connectorStart = { endpointNodeId: made[c.from].id, magnet: "AUTO" };
      conn.connectorEnd = { endpointNodeId: made[c.to].id, magnet: "AUTO" };
      conn.connectorStartStrokeCap = "NONE";
      conn.connectorEndStrokeCap = "ARROW_LINES";
      conn.strokes = [{ type: "SOLID", color: hex(CONNECTOR) }];
      conn.text.fontName = labelFont;
      conn.text.characters = c.label;
      connectorIds.push(conn.id);
      all.push(conn);
    }
    sectionIds.push(section.id);
    for (const s of Object.values(made)) stickyIds.push(s.id);
    all.push(section);
  }

  // Run card above the grid.
  let cardId: string | null = null;
  if (layout.sections.length) {
    const x0 = Math.min(...layout.sections.map((s) => s.x)) + dx;
    const y0 = Math.min(...layout.sections.map((s) => s.y)) + dy;
    const section = makeSection(`Motif · ${layout.card?.run_id ?? layout.run_id ?? "run"}`, "FFFFFF", x0, y0);
    const s = makeSticky(section, { key: "card", role: "card", wide: true, color: cardColor(layout), text: cardText(layout) });
    s.x = PAD; s.y = PAD;
    section.resizeWithoutConstraints(Math.max(layout.sections[0].width, s.width + 2 * PAD), s.height + 2 * PAD);
    section.y = y0 - section.height - CARD_GAP;
    cardId = section.id;
    stickyIds.push(s.id);
    all.push(section);
  }

  return finish(all, sectionIds, stickyIds, connectorIds, cardId, "figjam");
}

// ------------------------------------------------------------ Figma Design

async function renderDesign(layout: Layout, dx: number, dy: number): Promise<BoardResult> {
  const font: FontName = { family: "Inter", style: "Regular" };
  await figma.loadFontAsync(font);
  const sectionIds: string[] = [], stickyIds: string[] = [];
  const all: SceneNode[] = [];

  const makeSection = (name: string, fill: string, x: number, y: number): SectionNode => {
    const section = figma.createSection();
    section.name = name;
    section.fills = [{ type: "SOLID", color: hex(fill) }];
    section.x = x; section.y = y;
    return section;
  };
  const makeNote = (section: SectionNode, st: Sticky): FrameNode => {
    const f = figma.createFrame();
    f.name = st.key;
    f.layoutMode = "VERTICAL";
    f.paddingTop = f.paddingBottom = f.paddingLeft = f.paddingRight = 16;
    f.cornerRadius = 4;
    f.fills = [{ type: "SOLID", color: hex(st.color) }];
    section.appendChild(f);
    f.resize(st.wide ? 480 : 240, 100);
    f.primaryAxisSizingMode = "AUTO";
    f.counterAxisSizingMode = "FIXED";
    const t = figma.createText();
    t.fontName = font;
    t.fontSize = 14;
    t.characters = st.text;
    f.appendChild(t);
    t.layoutSizingHorizontal = "FILL";
    t.textAutoResize = "HEIGHT";
    return f;
  };

  for (const sec of layout.sections) {
    const section = makeSection(sec.name, sec.fill, sec.x + dx, sec.y + dy);
    const made: Record<string, FrameNode> = {};
    for (const st of sec.stickies) made[st.key] = makeNote(section, st);
    if (Object.keys(made).length !== sec.stickies.length) {
      throw new Error(`section ${sec.insight_id}: ${Object.keys(made).length} of ${sec.stickies.length} notes created`);
    }
    const rows = [1, 2].map((r) => sec.stickies.filter((s) => rowOf(s) === r).map((s) => made[s.key]));
    let y = PAD, maxRight = 0;
    for (const row of rows) {
      if (!row.length) continue;
      let x = PAD;
      for (const f of row) { f.x = x; f.y = y; x += f.width + SPACING; }
      maxRight = Math.max(maxRight, x - SPACING);
      y += Math.max(...row.map((f) => f.height)) + SPACING;
    }
    section.resizeWithoutConstraints(Math.max(sec.width, maxRight + PAD), y - SPACING + PAD);
    sectionIds.push(section.id);
    for (const f of Object.values(made)) stickyIds.push(f.id);
    all.push(section);
  }

  let cardId: string | null = null;
  if (layout.sections.length) {
    const x0 = Math.min(...layout.sections.map((s) => s.x)) + dx;
    const y0 = Math.min(...layout.sections.map((s) => s.y)) + dy;
    const section = makeSection(`Motif · ${layout.card?.run_id ?? layout.run_id ?? "run"}`, "FFFFFF", x0, y0);
    const f = makeNote(section, { key: "card", role: "card", wide: true, color: cardColor(layout), text: cardText(layout) });
    f.x = PAD; f.y = PAD;
    section.resizeWithoutConstraints(Math.max(layout.sections[0].width, f.width + 2 * PAD), f.height + 2 * PAD);
    section.y = y0 - section.height - CARD_GAP;
    cardId = section.id;
    stickyIds.push(f.id);
    all.push(section);
  }
  return finish(all, sectionIds, stickyIds, [], cardId, "figma");
}

function finish(all: SceneNode[], sectionIds: string[], stickyIds: string[], connectorIds: string[],
                cardId: string | null, editor: string): BoardResult {
  let x0 = Infinity, y0 = Infinity, x1 = -Infinity, y1 = -Infinity;
  for (const n of all) {
    if (!("x" in n) || n.type === "CONNECTOR") continue;  // connectors sit between their endpoints; sections bound the board
    x0 = Math.min(x0, n.x); y0 = Math.min(y0, n.y); x1 = Math.max(x1, n.x + n.width); y1 = Math.max(y1, n.y + n.height);
  }
  return {
    sectionIds, stickyIds, connectorIds, cardId,
    nodeIds: [...sectionIds, ...stickyIds, ...connectorIds, ...(cardId ? [cardId] : [])],
    counts: { sections: sectionIds.length, stickies: stickyIds.length, connectors: connectorIds.length },
    bounds: { x: x0, y: y0, width: x1 - x0, height: y1 - y0 }, editor,
  };
}

/** Render a layout on the current page. Chooses a free origin unless one is given. */
export async function renderBoard(layout: Layout, opts: { origin?: [number, number]; editor?: string } = {}): Promise<BoardResult> {
  if (!layout?.sections?.length) throw new Error("the layout has no sections; nothing to put on the board");
  const expected = layout.sections.reduce((n, s) => n + s.stickies.length, 0);
  if (!expected) throw new Error("the layout has no stickies; nothing to put on the board");
  const origin = opts.origin ?? chooseOrigin(figma.currentPage);
  const dx = origin[0] - layout.origin[0], dy = origin[1] - layout.origin[1];
  const editor = opts.editor ?? figma.editorType;
  const res = editor === "figjam" ? await renderFigJam(layout, dx, dy) : await renderDesign(layout, dx, dy);
  const got = res.counts.stickies - (res.cardId ? 1 : 0);
  if (got !== expected) throw new Error(`board incomplete: ${got} of ${expected} stickies were created`);
  return res;
}
